web-ng/src/app/components/map/map.component.ts

```diff
@@ -384,60 +386,113 @@ export class MapComponent implements AfterViewInit, OnDestroy {
     marker.setIcon(newIcon);
   }
 
-  private addGeoJsonToMap(feature: GeoJsonFeature) {
-    const addedFeatures = this.map.data.addGeoJson(feature.geoJson);
-    // Set property 'layerId' so that it knows what color to render later.
-    addedFeatures.forEach(f => {
-      f.setProperty('featureId', feature.id);
-      f.setProperty('layerId', feature.layerId);
+  private addGeoJsonFeature(
+    color: string | undefined,
+    layerName: string | undefined,
+    feature: GeoJsonFeature
+  ) {
+    const paths: google.maps.LatLng[][] = [];
+    const layer = new google.maps.Data();
+    layer.addGeoJson(feature.geoJson);
+    layer.forEach(f => {
+      if (f.getGeometry() instanceof google.maps.Data.Polygon) {
+        paths.push(
+          ...this.geoJsonPolygonToPaths(
+            f.getGeometry() as google.maps.Data.Polygon
+          )
+        );
+      }
+      if (f.getGeometry() instanceof google.maps.Data.MultiPolygon) {
+        (f.getGeometry() as google.maps.Data.MultiPolygon)
+          .getArray()
+          .forEach(polygon =>
+            paths.push(...this.geoJsonPolygonToPaths(polygon))
+          );
+      }
     });
+
+    this.addPolygonToMap(feature.id, color, layerName, paths);
   }
 
-  private addPolygonToMap(project: Project, feature: PolygonFeature) {
-    const color = project.layers.get(feature.layerId)?.color;
-    const vertices: google.maps.LatLng[] = [];
-    feature.polygonVertices.map(vertex => {
-      vertices.push(new google.maps.LatLng(vertex.latitude, vertex.longitude));
-    });
+  private geoJsonPolygonToPaths(
+    polygon: google.maps.Data.Polygon
+  ): google.maps.LatLng[][] {
+    const paths: google.maps.LatLng[][] = polygon
+      .getArray()
+      .map(linearRing => linearRing.getArray());
+    return paths;
+  }
 
+  private addPolygonFeature(
+    color: string | undefined,
+    layerName: string | undefined,
+    feature: PolygonFeature
+  ) {
+    const vertices: google.maps.LatLng[] = feature.polygonVertices.map(
+      vertex => new google.maps.LatLng(vertex.latitude, vertex.longitude)
+    );
+
+    this.addPolygonToMap(feature.id, color, layerName, [vertices]);
+  }
+
+  private addPolygonToMap(
+    featureId: string,
+    color: string | undefined,
+    layerName: string | undefined,
+    paths: google.maps.LatLng[][]
+  ) {
     const polygon = new google.maps.Polygon({
-      paths: vertices,
+      paths: paths,
       clickable: true,
       strokeColor: color,
       strokeOpacity: 1,
       strokeWeight: normalPolygonStrokeWeight,
       fillOpacity: 0,
+      map: this.map.googleMap,
     });
+    polygon.set('id', featureId);
+    polygon.set('color', color);
+    polygon.set('layerName', layerName);
+    polygon.addListener('click', (event: google.maps.PolyMouseEvent) => {
+      if (this.disableMapClicks) {
```

Comment:
There's a lot going on in here; please refactor handler and intersecting polygon detection into their own methods.